### PR TITLE
83 refactor frontend

### PR DESCRIPTION
**Pull request by Lamp-Nguyen**

### CODE REVIEW

File: new-chatbot-app/client/src/App.jsx [hunk 1]
```diff
@@ -15,156 +10,85 @@ import {
...
       setIsConnected(false);
       setAttemptedConnection(true);
     });
 
-    socketIo.on('connect_timeout', (timeout) => {
+    socket.on('connect_timeout', (timeout) => {
       console.error('Connection Timeout:', timeout);
       setIsConnected(false);
       setAttemptedConnection(true);
     });
```
Comment (dominhnhut01) [inline new-chatbot-app/client/src/App.jsx:75]: I think these socket.on should be in the SocketContextProvider, then SocketContextProvider would export the variables to notify other component if there is new incoming message or client disconnects.

File: new-chatbot-app/client/src/App.jsx [hunk 2]
```diff
@@ -15,156 +10,85 @@ import {
...
-      socketIo.off('disconnect');
-      socketIo.off('connect_error');
-      socketIo.off('connect_timeout');
+      socket.off('message');
+      socket.off('disconnect');
+      socket.off('connect_error');
+      socket.off('connect_timeout');
     };
-  }, [welcomeMessageShown, setMessages]);
+  }, [isConnected, setMessage]);
```
Comment (dominhnhut01) [inline new-chatbot-app/client/src/App.jsx:82]: I think this should be in SocketContextProvider too

File: new-chatbot-app/client/src/components/OfflineTicketWidget.jsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+          details: details,
+          ua: navigator.userAgent,
+        },
+        (responseMessage) => {
+          console.log(responseMessage);
+          setStep("initial");
+        }
+      );
+    }
+  };
```
Comment (dominhnhut01) [inline new-chatbot-app/client/src/components/OfflineTicketWidget.jsx:32]: I think this should be in the SocketContextProvider, too. You can do something like:
function socketSendMessage(message) {
 socket.emit(
        "createTicket",
        {
          question: question,
          email: email,
          name: name,
          details: details,
          ua: navigator.userAgent,
        },
        (responseMessage) => {
          console.log(responseMessage);
          setStep("initial");
        }
      );
      }
      and then export this function for other Component to use easily

**Review by dominhnhut01 — CHANGES_REQUESTED** [review]
I left some comments

**Review by TranHVLoc — APPROVED** [review]

File: new-chatbot-app/client/src/App.jsx (hunk 1 above)
Comment (Lamp-Nguyen) [reply new-chatbot-app/client/src/App.jsx]: I moved the whole useEffect to SocketContextProvider and it seems to be working

File: new-chatbot-app/client/src/context/SocketContextProvider.jsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      setAttemptedConnection(true);
+    });
+
+    return () => {
+      socket.off('message');
+      socket.off('disconnect');
+      socket.off('connect_error');
+      socket.off('connect_timeout');
+    };
+  }, [isConnected, mcv.setMessage]);
```
Comment (dominhnhut01) [inline new-chatbot-app/client/src/context/SocketContextProvider.jsx:71]: One of the dependency for the useEffect here is a function, not a value; it makes useEffect fire randomly? I should be a value, shouldn't it?

Comment (Lamp-Nguyen) [reply new-chatbot-app/client/src/context/SocketContextProvider.jsx]: I manged to solve it by changing the socket to use useRef, it seems that the socket context was being rendered every time the input message change so the socket kept being initialized, still don't know why ¯\_(ツ)_/¯, but know it should only be initialized once. Code looks messy but I'll fix that as we move on.

**Review by TranHVLoc — DISMISSED** [review]

File: new-chatbot-app/client/src/context/SocketContextProvider.jsx (hunk 4 above)
Comment (dominhnhut01) [reply new-chatbot-app/client/src/context/SocketContextProvider.jsx]: Great, looks good now

**Review by dominhnhut01 — APPROVED** [review]
lgtm!

File: new-chatbot-app/client/src/components/OfflineTicketWidget.jsx [hunk 5]
```diff
@@ -9,7 +9,7 @@ const OfflineTicketWidget = () => {
   const [name, setName] = useState('');
   const [details, setDetails] = useState('');
 
-  const { socketContextValues: scv } = useContext(SocketContext);
+  const { socketContextValues: socketContextValues } = useContext(SocketContext);
```
Comment (TranHVLoc) [inline new-chatbot-app/client/src/components/OfflineTicketWidget.jsx:12]: If you use type socketContextValues, I recommend changing the context variable to a different name (maybe a character change) for best convention

**Review by TranHVLoc — APPROVED** [review]
Look good to me! The only comment I left is for convention so you can change it in future PR

File: new-chatbot-app/client/src/components/OfflineTicketWidget.jsx (hunk 5 above)
Comment (Lamp-Nguyen) [reply new-chatbot-app/client/src/components/OfflineTicketWidget.jsx:12]: Yeah I was supposed to remove that, I must have missed that one while I was going over the code. Will fix it in the next PR.

**Pull request merged**